docs/rfcs/1018-server-stored-device-keys.md

```diff
@@ -0,0 +1,319 @@
+are always available when the device is under use. This is not strictly needed though since
+the device keys are only used in conjunction with the server (to authenticate to the server,
+sign new data, and encrypt/decrypt data to/from the server).
+
+This is the basis of the design changes discussed here:
+
+- The user&device keys are stored server-side...
+- ...encrypted by the device's local key.
+- They are lazily fetched from the server when needed and only kept in memory.
+
+The choice of local key to protect the device keys bundle may seem odd since the encrypted
+data are indeed not stored locally. However it makes perfect sense since by doing this we
+ensure the device still corresponds to a single physical machine (the one in possession of
+the local key) that is merely relying on the server as a storage facility.
+
+This is a similar approach to how are currently handled the workspace keys.
+
+While making acceptable the storage of the user&device keys on the server, this
+approach alone doesn't protect much against the risk of a stolen device: once a
+successful brute-force attack on a weak password has been achieved, the attacker
+can simply query the server for the user&device keys.
+
+However this pave the way for introducing MFA (multi-factor authentication) to reject
+suspicious access (e.g. from a different IP address).
+A typical example would be to systematically require TOTP (time-based one-time password)
+when login to Parsec.
+
+### 2.3 - Special case for vault-stored devices
+
+Unlike other devices, registration devices (see [RFC 1015](1015-registration-device.md)) are not locally stored.
+
+Those devices are stored server-side in the Parsec account vault. Hence in this
+case the device keys are still stored in the local device since:
+
+- It avoid a round trip to the server to fetch the device keys bundle.
+- We have to support the case of keys in the local device for legacy device keys files anyway.
+
+> [!NOTE]
+> In case of web, while still stored locally, the local device is encrypted
+> with a high entropy key protected by the vault key and stored server-side (see [RFC 1016](1016-local-device-storage-in-web-client.md)).
+> So in this case we might also leave the device keys in the local device.
+
+## 4 - Datamodel
+
+### 4.1 - Option A: Update `LocalDevice`
+
+Changes:
+
+- Turn `signing_key` fields optional
+- Turn `private_key` fields optional
+- Add optional `keys_bundle_token` field
+- Add validation rule to ensure `signing_key`&`private_key` are both set or
+  `keys_bundle_token` is set.
+
+### 4.2 - (not chosen) Option B: New `DeviceLocalBundle`
+
+Option A has the advantage of being backward compatible with the current, however it
+doesn't offer a clean separation of `signing_key`&`private_key` on one side and
+`keys_bundle_token` on the other side.
+
+Hence option B that instead considers `LocalDevice` as a legacy type (only used for
+device keys files).
+
+```json5
```

Comment:
nit: I'm not sure we want to keep this with such details as it is not the chosen option

Reply:
I guess that at this stage, it does not hurt to leave it like that. It gives some context and it is already specified that it is not the chosen option.